Add HSTU in fbgemm_gpu/experimental/

jiayus-nvidia opened the pull request:
Summary: A cutlass based HSTU for both Ampere and Hopper, both forward and backward is added to fbgemm_gpu/experimental/.

jiayus-nvidia committed:
Add support for HSTU in fbgemm_gpu/experimental/hstu.

q10 committed:
Remove `sm_100` and `sm_120` (#4024)

Summary:
X-link: https://github.com/facebookresearch/FBGEMM/pull/1133

- Remove sm_100 and sm_120 from architectures list and keep just sm_100a and sm_120a instead, to enable compilation for FP4 CUTLASS quantization kernels (https://github.com/pytorch/FBGEMM/pull/4004), since we are running into the following error:

```
Instruction 'cvt with .e2m1x2' not supported on .target 'sm_100'
```

Pull Request resolved: https://github.com/pytorch/FBGEMM/pull/4024

Reviewed By: spcyppt

Differential Revision: D73901832

Pulled By: q10

fbshipit-source-id: 690c58b214aee80374e43a93bf39fe70e430da9a

jiayus-nvidia committed:
Merge branch 'pytorch:main' into main

jiayus-nvidia committed:
Delete fbgemm_gpu/experimental/hstu/test/test.py

jiayus-nvidia committed:
Merge branch 'main' into main

jiayus-nvidia committed:
Update mainloop_bwd_sm90_tma_gmma_ws.hpp

jiayus-nvidia committed:
Merge branch 'pytorch:main' into main

jiayus-nvidia committed:
Merge branch 'main' into main

jiayus-nvidia committed:
Fix lint errors

jiayus-nvidia committed:
Merge branch 'pytorch:main' into main

jwfromm commented:
This is fantastic @jiayus-nvidia! We really appreciate your contribution. Just for my understanding, do you have any benchmarking information on these new kernels you can share?

jiayus-nvidia commented:
Hi @jwfromm, below are some benchmarking information. The first four figures are fwd and bwd kernels with rab (and drab), compared with some commit in [generative-recommenders](https://github.com/meta-recsys/generative-recommenders/tree/main) around last November, so the performance for triton kernel must improve a lot by now. And the last two figures are fwd and bwd kernels without rab. I haven't had the chance to measure the performance of the latest version of the triton kernel yet, so I just provide the performance of our kernel for reference first. If you need performance data for more dimensions or sequence lengths, feel free to let me know.

![A100_rab_bwd](https://github.com/user-attachments/assets/eeab70ca-bc1e-4e6b-8574-eb2c1ab45bb6)
![A100_rab_fwd](https://github.com/user-attachments/assets/992ccef1-cf4d-49bd-a4b5-eaa5b41d0425)
![H100_rab_bwd](https://github.com/user-attachments/assets/bf9b24f1-ae01-4abc-9f80-8c8129cabd85)
![H100_rab_fwd](https://github.com/user-attachments/assets/34b5fe8b-27e6-4dc0-a968-52ff0a35a120)
![A100_no_rab_fwd_bwd](https://github.com/user-attachments/assets/9c1a3f77-d7c8-4f26-8170-73226b3163e1)
![H100_no_rab_fwd_bwd](https://github.com/user-attachments/assets/fdb1a58a-c53c-4971-b60a-5da4c1657a43)

jiayus-nvidia committed:
Performance optimization and lint fixes

jiayus-nvidia committed:
Merge branch 'main' of https://github.com/jiayus-nvidia/FBGEMM

Commit:
Fixed the HSTU lint errors

Summary: Ran linters + fixed issues

Differential Revision: D75262862

ionuthristodorescu commented:
Hi @jiayus-nvidia - thank you for the PR and the benchmark results, I am verifying it internally and I see some lints, formatting errors and some clang warnings about "CUDA indexes declared as int". I will send you a diff with the fixes for your review as soon as the export-to-Github finishes. Thank you again !

jiayus-nvidia committed:
Merge branch 'pytorch:main' into main

Commit:
Fixing the linter errors in the HSTU import

Summary: There are some leftover linter errors in the HSTU import, this diff fixes them.

Reviewed By: q10

Differential Revision: D75545512

ionuthristodorescu commented:
> Hi @ionuthristodorescu , thanks for your help! I have merged the commit into this PR.

Hi @jiayus-nvidia - I re-ran the internal CI/CD and looks like there were some new lint errors in the new code, I pushed the fixes in https://github.com/pytorch/FBGEMM/pull/4218/commits/f5e63b5b401b1d43c76f0e5e02faf76850b0e06a (second commit in #4218). Could you cherry-pick them and push to this PR again (this time without any new code so no new lint errors will be triggered) - I can re-import and merge after that. Thank you !

Commit:
[lintfix] newline at EOF in fbgemm_gpu/experimental/hstu/CMakeLists.txt

rgommers commented on fbgemm_gpu/experimental/hstu/LICENSE:
```diff
@@ -0,0 +1,41 @@
+
+/*
+ * SPDX-FileCopyrightText: Copyright (c) <2024> NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: LicenseRef-NvidiaProprietary
```
This isn't valid content for a license file and looks like a copy-paste error that was missed. If it's real, surely that would be against the CLA. I opened issue gh-5757 for this.
